File: ios/Classes/FlutterUgcUploadPlugin.m
#import "FlutterUgcUploadPlugin.h"
#import "TXUGCPublish.h"

@implementation FlutterUgcUploadPlugin{
    FlutterEventSink _eventSink;
    TXUGCPublish *_txUgcPublish;
    NSString *customKey;
}
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"flutter_ugc_upload"
            binaryMessenger:[registrar messenger]];
  FlutterUgcUploadPlugin* instance = [[FlutterUgcUploadPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
    
    FlutterEventChannel *eventChannel = [FlutterEventChannel eventChannelWithName:@"flutter_ugc_upload_stream" binaryMessenger:[registrar messenger]];
    [eventChannel setStreamHandler:instance];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  }else if ([@"setCustomKey" isEqualToString:call.method]) {
      [self setCustomKey:call result:result];
  } else if ([@"uploadVideo" isEqualToString:call.method]){
      [self uploadVideo : call result:result];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)setCustomKey:(FlutterMethodCall*)call
              result:(FlutterResult)result
{
    NSString *customKey = call.arguments[@"customKey"];
    
    if (self->_txUgcPublish == nil || (customKey != nil && ![customKey isEqualToString:self->customKey])) {
        self->_txUgcPublish = [[TXUGCPublish alloc] initWithUserID:customKey];
        self->_txUgcPublish.delegate = (id)self;
    }

    self->customKey = customKey;
}


- (void) uploadVideo:(FlutterMethodCall*)call
              result:(FlutterResult)result {
    NSString *signature = call.arguments[@"signature"];
    NSString *videoPath = call.arguments[@"videoPath"];
    NSString *coverPath = call.arguments[@"coverPath"];
    
    TXPublishParam *publishParam = [[TXPublishParam alloc] init];
    publishParam.enableHTTPS  = true;
    publishParam.signature  = signature;
    publishParam.videoPath  = videoPath;
    if(![coverPath isEqual:@""]){
        publishParam.coverPath  = coverPath;
    }
    int ret = [_txUgcPublish publishVideo:publishParam];
    result([NSNumber numberWithInt:ret]);
    
}

#pragma mark - TXVideoPublishListener

- (void)onPublishProgress:(NSInteger)uploadBytes totalBytes:(NSInteger)totalBytes {
    //self.progressView.progress = (float)uploadBytes/totalBytes;
    NSDictionary<NSString *, id> *eventData = @{
                   @"uploadBytes": @(uploadBytes),
                   @"totalBytes": @(totalBytes),
                   @"method": @"publishProgress"
           
       };
    self->_eventSink(eventData);
}

- (void)onPublishComplete:(TXPublishResult*)res {
    //NSString *string = [NSString stringWithFormat:@"上传完成，错误码[%d]，信息[%@]", res.retCode, res.retCode == 0? res.videoURL: res.descMsg];
    NSDictionary<NSString *, id> *eventData =  @{
                           @"retCode": @(res.retCode),
                           @"descMsg": res.descMsg != nil ? res.descMsg : @"",
                           @"videoId": res.videoId != nil ? res.videoId : @"",
                           @"videoURL": res.videoURL != nil ? res.videoURL : @"",
                           @"coverURL": res.coverURL != nil ? res.coverURL : @"",
                           @"method": @"publishComplete"
       };
    self->_eventSink(eventData);
}

- (FlutterError * _Nullable)onCancelWithArguments:(id _Nullable)arguments {
    _eventSink = nil;
    return nil;
}

- (FlutterError * _Nullable)onListenWithArguments:(id _Nullable)arguments eventSink:(nonnull FlutterEventSink)events {
    _eventSink = events;
    return nil;
}

@end

File: android/src/main/java/com/example/miao/flutter_ugc_upload/FlutterUgcUploadPlugin.java
package com.example.miao.flutter_ugc_upload;

import android.content.Context;

import androidx.annotation.NonNull;

import com.example.miao.flutter_ugc_upload.videoupload.TXUGCPublish;
import com.example.miao.flutter_ugc_upload.videoupload.TXUGCPublishTypeDef;

import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** FlutterUgcUploadPlugin */
public class FlutterUgcUploadPlugin implements FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

  private MethodChannel channel;
  private static Context mContext;
  private EventChannel eventChannel;
  public static EventChannel.EventSink mEventSink = null;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    if (null == mContext) {
      mContext = flutterPluginBinding.getApplicationContext();
    }
    /**
     * 方法调用通道
     */
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_ugc_upload");
    channel.setMethodCallHandler(this);
    /**
     * 回调监听通道
     */
    eventChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_ugc_upload_stream");
    eventChannel.setStreamHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else if(call.method.equals("uploadVideo")){
      uploadVideo(call,result);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

  private void uploadVideo(MethodCall call, Result result){
    TXUGCPublish mVideoPublish = new TXUGCPublish(mContext, "independence_android");
    mVideoPublish.setListener(new TXUGCPublishTypeDef.ITXVideoPublishListener() {
      @Override
      public void onPublishProgress(long uploadBytes, long totalBytes) {
        Map map =new HashMap();
        map.put("uploadBytes", uploadBytes);
        map.put("totalBytes",  totalBytes);
        map.put("method",  "publishProgress");
        mEventSink.success(map);
      }

      @Override
      public void onPublishComplete(TXUGCPublishTypeDef.TXPublishResult res) {
        Map map = new HashMap();
        map.put("method",  "publishComplete");
        map.put("retCode",res.retCode);
        map.put("descMsg",res.descMsg);
        map.put("videoId",res.videoId);
        map.put("videoURL",res.videoURL);
        map.put("coverURL",res.coverURL);
        mEventSink.success(map);
      }
    });
    TXUGCPublishTypeDef.TXPublishParam param = new TXUGCPublishTypeDef.TXPublishParam();
    param.enableHttps = true;
    param.signature = call.argument("signature");
    param.videoPath = call.argument("videoPath");
    if(call.argument("coverPath")!= "") {
      param.coverPath = call.argument("coverPath");
    }
    int publishCode = mVideoPublish.publishVideo(param);
    result.success(publishCode);
  }

  @Override
  public void onListen(Object arguments, EventChannel.EventSink events) {
    mEventSink = events;
  }

  @Override
  public void onCancel(Object arguments) {
    mEventSink.endOfStream();
  }
}
